lib/services/database/challenge_repository_service.dart

```diff
@@ -0,0 +1,222 @@
+import "dart:core";
+
+import "package:cloud_firestore/cloud_firestore.dart";
+import "package:hooks_riverpod/hooks_riverpod.dart";
+import "package:proxima/models/database/challenge/challenge_data.dart";
+import "package:proxima/models/database/challenge/challenge_firestore.dart";
+import "package:proxima/models/database/post/post_firestore.dart";
+import "package:proxima/models/database/post/post_id_firestore.dart";
+import "package:proxima/models/database/user/user_firestore.dart";
+import "package:proxima/models/database/user/user_id_firestore.dart";
+import "package:proxima/services/database/firestore_service.dart";
+import "package:proxima/services/database/post_repository_service.dart";
+
+/// This repository service is responsible for handling the challenges
+class ChallengeRepositoryService {
+  final FirebaseFirestore _firestore;
+  final PostRepositoryService _postRepositoryService;
+
+  static const int maxActiveChallenges = 3;
+  static const double maxChallengeRadius = 3; // in km
+  static const double minChallengeRadius = 0.5;
+  static const maxChallengeDuration = Duration(days: 1);
+
+  /// Creates a new challenge repository service
+  /// with the given [firestore] and [postRepositoryService]
+  ChallengeRepositoryService({
+    required FirebaseFirestore firestore,
+    required PostRepositoryService postRepositoryService,
+  })  : _firestore = firestore,
+        _postRepositoryService = postRepositoryService;
+
+  CollectionReference<Map<String, dynamic>> _activeChallengesRef(
+    DocumentReference parentRef,
+  ) {
+    return parentRef.collection(ChallengeFirestore.subCollectionName);
+  }
+
+  CollectionReference<Map<String, dynamic>> _pastChallengesRef(
+    DocumentReference parentRef,
+  ) {
+    return parentRef
+        .collection(ChallengeFirestore.pastChallengesSubCollectionName);
+  }
+
+  /// Completes the challenge of the user with id [uid] on the post with id [pid]
+  Future<void> completeChallenge(
+    UserIdFirestore uid,
+    PostIdFirestore pid,
+  ) async {
+    final userDocRef =
+        _firestore.collection(UserFirestore.collectionName).doc(uid.value);
+    await _completeChallenge(userDocRef, pid);
+  }
+
+  Future<void> _completeChallenge(
```

Comment:
By humble curiosity, what is the purpose of this private method? Why isn't it inlined?

Reply:
The idea is to make it simple to reuse with group documents later on.

Reply:
But this will be another repository anyway, won't it? We will have a repository for handling votes on comments, and one for handling votes for posts; why would it be different here?

Reply:
Alright I merged the private generic `_getChallenges` and `_completeChallenge` methods with the public user ones. The helper private methods still think in terms of parent refs but I think that's completely fine and does not decrease readability. c5e5099aa080cf9ec1d1fb0433ad9f24b3a54821